plugin/trino-opensearch/src/main/java/io/trino/plugin/opensearch/OpenSearchMetadata.java

```diff
@@ -660,6 +688,174 @@ private static boolean isPassthroughQuery(OpenSearchTableHandle table)
+            String projectedColumnName = projectedColumnHandle.getQualifiedName();
+
+            Variable projectedColumnVariable = new Variable(projectedColumnName, expression.getType());
+            Assignment newAssignment = new Assignment(projectedColumnName, projectedColumnHandle, expression.getType());
+            newAssignments.putIfAbsent(projectedColumnName, newAssignment);
+
+            newVariablesBuilder.put(expression, projectedColumnVariable);
+            projectedColumnsBuilder.add(projectedColumnHandle);
+        }
+
+        // Modify projections to refer to new variables
+        Map<ConnectorExpression, Variable> newVariables = newVariablesBuilder.buildOrThrow();
+        List<ConnectorExpression> newProjections = projections.stream()
+                .map(expression -> replaceWithNewVariables(expression, newVariables))
+                .collect(toImmutableList());
+
+        List<Assignment> outputAssignments = newAssignments.values().stream().collect(toImmutableList());
+        return Optional.of(new ProjectionApplicationResult<>(
+                openSearchTableHandle.withProjectedColumns(projectedColumnsBuilder.build()),
+                newProjections,
+                outputAssignments,
+                false));
+    }
+
+    private static boolean containsCallToJsonCastFunction(List<ConnectorExpression> projections)
+    {
+        List<Call> callProjections = projections.stream().filter(call -> call instanceof Call).map(call -> (Call) call).toList();
+        if (callProjections.isEmpty()) {
+            return false;
+        }
+        boolean isCastFunction = false;
+        for (Call call : callProjections) {
+            if (containsCallToJsonCastFunction(call)) {
+                return true;
+            }
+            isCastFunction = containsCallToJsonCastFunction(call.getArguments()) || isCastFunction;
+        }
+        return isCastFunction;
+    }
+
+    private static boolean containsCallToJsonCastFunction(Call call)
+    {
+        return switch (call.getFunctionName().getName()) {
+            case "$internal$json_string_to_array_cast", "$internal$json_string_to_row_cast", "$internal$json_string_to_map_cast" -> true;
+            default -> false;
+        };
+    }
+
+    private static boolean isSupportedForPushdown(ConnectorExpression connectorExpression)
+    {
+        if (connectorExpression instanceof Variable) {
+            return true;
+        }
+        if (connectorExpression instanceof FieldDereference fieldDereference) {
+            RowType rowType = (RowType) fieldDereference.getTarget().getType();
+            RowType.Field field = rowType.getFields().get(fieldDereference.getField());
+            if (field.getName().isEmpty()) {
+                return false;
+            }
+            return !field.getName().get().contains(".");
```

Comment:
Does this connector allow fields with dots?

Reply:
Yes not needed. Removed and refactored.